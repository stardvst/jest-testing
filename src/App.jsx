//import "./App.css";

import DataFetch from "./DataFetch";
import React from "react";
import axios from "axios";
import { hot } from "react-hot-loader";

export const dataReducer = (state, action) => {
  if (action.type === "SET_ERROR") {
    return { ...state, list: [], error: true };
  }

  if (action.type === "SET_LIST") {
    return { ...state, list: action.list, error: null };
  }

  throw new Error();
};

const initialData = {
  list: [],
  error: null,
};

const getUser = () => {
  return Promise.resolve({ id: 1, name: "Robin" });
};

const App = () => {
  const [counter, setCounter] = React.useState(0);
  const [data, dispatch] = React.useReducer(dataReducer, initialData);
  const [user, setUser] = React.useState(null);

  React.useEffect(() => {
    axios
      .get("http://hn.algolia.com/api/v1/search?query=react")
      .then((response) => {
        dispatch({ type: "SET_LIST", list: response.data.hits });
      })
      .catch(() => {
        dispatch({ type: "SET_ERROR" });
      });

    (async () => {
      const user = await getUser();
      setUser(user);
    })();
  }, []);

  return (
    <div>
      <h1>{user ? `${user.name}s Counter` : ""}</h1>
      <Counter counter={counter} />

      <button type="button" onClick={() => setCounter(counter + 1)}>
        Increment
      </button>

      <button type="button" onClick={() => setCounter(counter - 1)}>
        Decrement
      </button>

      <h2>My Async Data</h2>

      {data.error && <div className="error">Error</div>}

      <ul>
        {data.list.map((item) => (
          <li key={item.objectID}>{item.title}</li>
        ))}
      </ul>
    </div>
  );
};

export const Counter = ({ counter, onChange = null }) => (
  <div>
    <p>{counter}</p>
    <input type="text" onChange={onChange} />
  </div>
);

export const sum = (a, b) => a + b;

export default hot(module)(App);
