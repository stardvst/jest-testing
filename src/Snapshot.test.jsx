import App, { Counter, dataReducer, sum } from "./App";

import React from "react";
import renderer from "react-test-renderer";

const list = ["a", "b", "c"];

describe("App", () => {
  describe("Reducer", () => {
    it("should set a list", () => {
      const state = { list: [], error: null };
      const newState = dataReducer(state, {
        type: "SET_LIST",
        list,
      });

      expect(newState).toEqual({ list, error: null });
    });

    it("should reset the error if list is set", () => {
      const state = { list: [], error: true };
      const newState = dataReducer(state, {
        type: "SET_LIST",
        list,
      });

      expect(newState).toEqual({ list, error: null });
    });

    it("should set the error", () => {
      const state = { list: [], error: null };
      const newState = dataReducer(state, {
        type: "SET_ERROR",
      });

      expect(newState.error).toBeTruthy();
    });
  });

  test("snapshot renders", () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Counter", () => {
  test("snapshot renders", () => {
    const component = renderer.create(<Counter counter={1} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("sum", () => {
  test("should sum two values", () => {
    expect(sum(2, 4)).toBe(6);
  });
});
