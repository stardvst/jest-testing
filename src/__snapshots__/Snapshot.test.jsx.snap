// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
<div>
  <h1>
    
  </h1>
  <div>
    <p>
      0
    </p>
    <input
      type="text"
    />
  </div>
  <button
    onClick={[Function]}
    type="button"
  >
    Increment
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Decrement
  </button>
  <h2>
    My Async Data
  </h2>
  <ul />
</div>
`;

exports[`Counter snapshot renders 1`] = `
<div>
  <p>
    1
  </p>
  <input
    type="text"
  />
</div>
`;
